spellbook/optimization/qoi.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+import numpy as np
+
+from spellbook.utils import load_infile
+
+
+""" Construct a quantity of interest (cost function) for use in optimization"""
+
+
+def barrier(x, threshold, threshold_type="greater"):
+    """Implement the 'barrier penalty' function from Haftka and Starnes, AIAA Journal 14 (6): 718 (1976).
+
+    x is the quantity
+    threshold is the constraint threshold
+    threshold_type: one of "greater" or "less"
+
+    Note: this is just the penalty term, which must be combined with the objective
+    Use this with make_barrier_qoi to apply them
+
+    """
```

Comment:
It might be useful to add a couple unit tests to check the returned penalty makes sense.

```python
# generate 10 random numbers on [-1, 1]
x = 1.0 + 2.0*np.random(10)
threshold = 0.01

penalties = barrier(x, threshold, "greater")
for i, penalty in enumerate(penalties):
    if x[i] > threshold:
        assert penalty > 0
    else:
        assert penalty >= 0

penalties = barrier(x, threshold, "less")
for i, penalty in enumerate(penalties):
    if x[i] < threshold:
        assert penalty > 0
    else:
        assert penalty >= 0
```

I guess you could also check for a negative threshold too.